gateway/src/endpoints/openai_compatible.rs

```diff
@@ -0,0 +1,1020 @@
+        };
+        let variant_name = headers
+            .get("variant_name")
+            .map(|h| {
+                h.to_str()
+                    .map_err(|_| Error::InvalidOpenAICompatibleRequest {
+                        message: "variant_name header is not valid UTF-8".to_string(),
+                    })
+                    .map(|s| s.to_string())
+            })
+            .transpose()?;
+        let dryrun = headers
+            .get("dryrun")
+            .map(|h| {
+                h.to_str()
+                    .map_err(|_| Error::InvalidOpenAICompatibleRequest {
+                        message: "dryrun header is not valid UTF-8".to_string(),
+                    })
+                    .and_then(|s| {
+                        s.parse::<bool>()
+                            .map_err(|_| Error::InvalidOpenAICompatibleRequest {
+                                message: "dryrun header is not a valid boolean".to_string(),
+                            })
+                    })
+            })
+            .transpose()?;
+        let dynamic_tool_params = DynamicToolParams {
+            allowed_tools: None,
+            additional_tools: openai_compatible_params
+                .tools
+                .map(|tools| tools.into_iter().map(|tool| tool.into()).collect()),
+            tool_choice: openai_compatible_params
+                .tool_choice
+                .map(|tool_choice| tool_choice.into()),
+            parallel_tool_calls: openai_compatible_params.parallel_tool_calls,
+        };
+        let output_schema = match openai_compatible_params.response_format {
+            Some(OpenAICompatibleResponseFormat::JsonSchema { schema }) => Some(schema),
+            _ => None,
+        };
+        Ok(Params {
+            function_name,
+            episode_id,
+            input,
+            stream: openai_compatible_params.stream,
+            params: inference_params,
+            variant_name,
+            dryrun,
+            dynamic_tool_params,
+            output_schema,
+            credentials: InferenceCredentials::default(),
+        })
+    }
+}
+
+impl TryFrom<Vec<OpenAICompatibleMessage>> for Input {
+    type Error = Error;
+    fn try_from(
+        openai_compatible_messages: Vec<OpenAICompatibleMessage>,
+    ) -> Result<Self, Self::Error> {
+        let mut system = None;
+        let mut messages = Vec::new();
+        let mut tool_call_id_to_name = HashMap::new();
+        for message in openai_compatible_messages {
+            match message {
+                OpenAICompatibleMessage::System(msg) => {
+                    if system.is_some() {
+                        return Err(Error::InvalidOpenAICompatibleRequest {
+                            message: "At most one system message is allowed".to_string(),
+                        });
+                    }
+                    system = Some(convert_openai_message_content(msg.content)?);
```

Comment:
This should explicitly check it's the first, or fail.

Reply:
I dont' see why we can't do best-effort with that

Reply:
we should not reorder messages without the user knowing